server/channels/app/session.go

```diff
@@ -291,6 +291,30 @@ func (a *App) AttachDeviceId(sessionID string, deviceID string, expiresAt int64)
 	return nil
 }
 
+func (a *App) SetExtraSessionProps(session *model.Session, newProps map[string]string) *model.AppError {
+	changed := false
+	for k, v := range newProps {
+		if session.Props[k] == v {
+			continue
+		}
+
+		session.AddProp(k, v)
+		changed = true
+	}
+
+	if !changed {
+		return nil
+	}
+
+	err := a.Srv().Store().Session().UpdateProps(session)
+	a.Srv().Platform().AddSessionToCache(session)
+	if err != nil {
+		return model.NewAppError("SetExtraSessionProps", "app.session.set_extra_session_prop.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+
+	return nil
+}
```

Comment:
(Not sure why this was not posted before)

Shouldn't we update the cache only if the update didn't return an error? 

```suggestion
	if err := a.Srv().Store().Session().UpdateProps(session); err != nil {
		return model.NewAppError("SetExtraSessionProps", "app.session.set_extra_session_prop.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
	}

	a.Srv().Platform().AddSessionToCache(session)

	return nil
}
```